File: app/src/main/java/com/raju/disney/opentelemetry/Pre29ActivityCallbacks.java
package com.raju.disney.opentelemetry;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import io.opentelemetry.api.trace.Tracer;

class Pre29ActivityCallbacks implements Application.ActivityLifecycleCallbacks {
    private final Tracer tracer;
    private final VisibleScreenTracker visibleScreenTracker;
    private final Map<String, ActivityTracer> tracersByActivityClassName = new HashMap<>();
    private final AtomicReference<String> initialAppActivity = new AtomicReference<>();
    private final AppStartupTimer appStartupTimer;
    private final List<AppStateListener> appStateListeners;
    private int numberOfOpenActivities = 0;

    Pre29ActivityCallbacks(Tracer tracer, VisibleScreenTracker visibleScreenTracker, AppStartupTimer appStartupTimer, List<AppStateListener> appStateListeners) {
        this.tracer = tracer;
        this.visibleScreenTracker = visibleScreenTracker;
        this.appStartupTimer = appStartupTimer;
        this.appStateListeners = appStateListeners;
    }

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
        getTracer(activity)
                .startActivityCreation()
                .addEvent("activityCreated");

        if (activity instanceof FragmentActivity) {
            FragmentManager fragmentManager = ((FragmentActivity) activity).getSupportFragmentManager();
            fragmentManager.registerFragmentLifecycleCallbacks(new OtelFragmentLifecycleCallbacks(tracer, visibleScreenTracker), true);
        }
    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {
        if (numberOfOpenActivities == 0) {
            for (AppStateListener appListener : appStateListeners) {
                appListener.appForegrounded();
            }
        }
        numberOfOpenActivities++;
        getTracer(activity)
                .initiateRestartSpanIfNecessary(tracersByActivityClassName.size() > 1)
                .addEvent("activityStarted");
    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {
        getTracer(activity)
                .startSpanIfNoneInProgress("Resumed:" + activity.getClass().getSimpleName())
                .addEvent("activityResumed")
                .addPreviousScreenAttribute()
                .endSpanForActivityResumed();
        visibleScreenTracker.activityResumed(activity);
    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {
        getTracer(activity)
                .startSpanIfNoneInProgress("Paused:" + activity.getClass().getSimpleName())
                .addEvent("activityPaused")
                .endActiveSpan();
        visibleScreenTracker.activityPaused(activity);
    }

    @Override
    public void onActivityStopped(@NonNull Activity activity) {
        if (--numberOfOpenActivities == 0) {
            for (AppStateListener appListener : appStateListeners) {
                appListener.appBackgrounded();
            }
        }
        getTracer(activity)
                .startSpanIfNoneInProgress("Stopped:" + activity.getClass().getSimpleName())
                .addEvent("activityStopped")
                .endActiveSpan();
    }

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {
        //todo: add event
    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {
        getTracer(activity)
                .startSpanIfNoneInProgress("Destroyed:" + activity.getClass().getSimpleName())
                .addEvent("activityDestroyed")
                .endActiveSpan();
    }

    private ActivityTracer getTracer(Activity activity) {
        ActivityTracer activityTracer = tracersByActivityClassName.get(activity.getClass().getName());
        if (activityTracer == null) {
            activityTracer = new ActivityTracer(activity, initialAppActivity, tracer, visibleScreenTracker, appStartupTimer);
            tracersByActivityClassName.put(activity.getClass().getName(), activityTracer);
        }
        return activityTracer;
    }

}

File: app/src/main/java/com/raju/disney/opentelemetry/Pre29ActivityCallbacks.kt
package com.raju.disney.opentelemetry

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import io.opentelemetry.api.trace.Tracer
import java.util.concurrent.atomic.AtomicReference

internal class Pre29ActivityCallbacks(
    private val tracer: Tracer,
    private val visibleScreenTracker: VisibleScreenTracker,
    private val appStartupTimer: AppStartupTimer,
    private val appStateListeners: List<AppStateListener>
) : ActivityLifecycleCallbacks {

    private val tracersByActivityClassName: MutableMap<String, ActivityTracer> = HashMap()
    private val initialAppActivity = AtomicReference<String>()
    private var numberOfOpenActivities = 0

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        getTracer(activity)
            .startActivityCreation()
            .addEvent("activityCreated")
        if (activity is FragmentActivity) {
            val fragmentManager = activity.supportFragmentManager
            fragmentManager.registerFragmentLifecycleCallbacks(
                FragmentLifecycleCallbacks(
                    tracer,
                    visibleScreenTracker
                ), true
            )
        }
    }

    override fun onActivityStarted(activity: Activity) {
        if (numberOfOpenActivities == 0) {
            for (appListener in appStateListeners) {
                appListener.appForegrounded()
            }
        }
        numberOfOpenActivities++
        getTracer(activity)
            .initiateRestartSpanIfNecessary(tracersByActivityClassName.size > 1)
            .addEvent("activityStarted")
    }

    override fun onActivityResumed(activity: Activity) {
        getTracer(activity)
            .startSpanIfNoneInProgress("Resumed:" + activity.javaClass.simpleName)
            .addEvent("activityResumed")
            .addPreviousScreenAttribute()
            .endSpanForActivityResumed()
        visibleScreenTracker.activityResumed(activity)
    }

    override fun onActivityPaused(activity: Activity) {
        getTracer(activity)
            .startSpanIfNoneInProgress("Paused:" + activity.javaClass.simpleName)
            .addEvent("activityPaused")
            .endActiveSpan()
        visibleScreenTracker.activityPaused(activity)
    }

    override fun onActivityStopped(activity: Activity) {
        if (--numberOfOpenActivities == 0) {
            for (appListener in appStateListeners) {
                appListener.appBackgrounded()
            }
        }
        getTracer(activity)
            .startSpanIfNoneInProgress("Stopped:" + activity.javaClass.simpleName)
            .addEvent("activityStopped")
            .endActiveSpan()
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        //todo: add event
    }

    override fun onActivityDestroyed(activity: Activity) {
        getTracer(activity)
            .startSpanIfNoneInProgress("Destroyed:" + activity.javaClass.simpleName)
            .addEvent("activityDestroyed")
            .endActiveSpan()
    }

    private fun getTracer(activity: Activity): ActivityTracer {
        var activityTracer = tracersByActivityClassName[activity.javaClass.name]
        if (activityTracer == null) {
            activityTracer = ActivityTracer(
                activity,
                initialAppActivity,
                tracer,
                visibleScreenTracker,
                appStartupTimer
            )
            tracersByActivityClassName[activity.javaClass.name] = activityTracer
        }
        return activityTracer
    }
}